kyaml/kio/filters/merge3.go

```diff
@@ -110,10 +115,14 @@ func (ts *tuples) add(node *yaml.RNode) error {
 	if err != nil {
 		return err
 	}
+	nodePath := nodeMeta.Annotations[kioutil.PathAnnotation]
 	for i := range ts.list {
 		t := ts.list[i]
+
+		path := t.meta.Annotations[kioutil.PathAnnotation]
 		if t.meta.Name == nodeMeta.Name && t.meta.Namespace == nodeMeta.Namespace &&
-			t.meta.APIVersion == nodeMeta.APIVersion && t.meta.Kind == nodeMeta.Kind {
+			t.meta.APIVersion == nodeMeta.APIVersion && t.meta.Kind == nodeMeta.Kind &&
+			(!ts.mergeOnPath || nodePath == path) {
```

Comment:
would be good to add a comment somewhere explaining `why` part of this logic here.

Reply:
PTAL